File: mapping/legacy/link_rooms.py
#!/usr/bin/env python3
# link_rooms.py
# Links parsed room JSONs by movement commands in a GemStone-style log.

import re
import json
import hashlib
import argparse

ORDINALS = [
    "north", "south", "east", "west",
    "northeast", "southeast", "northwest", "southwest",
    "up", "down", "out",
    # Abbreviations
    "n", "s", "e", "w", "ne", "se", "nw", "sw", "u", "d"
]

# Map abbreviations to full directions for reverse lookup
ABBREVIATIONS = {
    "n": "north", "s": "south", "e": "east", "w": "west",
    "ne": "northeast", "se": "southeast", "nw": "northwest", "sw": "southwest",
    "u": "up", "d": "down"
}

REVERSE = {
    "north": "south", "south": "north",
    "east": "west", "west": "east",
    "northeast": "southwest", "southwest": "northeast",
    "northwest": "southeast", "southeast": "northwest",
    "up": "down", "down": "up", "out": "out",
    # Abbreviations
    "n": "south", "s": "north", "e": "west", "w": "east",
    "ne": "southwest", "se": "northwest", "nw": "southeast", "sw": "northeast",
    "u": "down", "d": "up"
}

HEADER_RE = re.compile(r'^\[(?P<title>.+?)\]\s*$')

def link_rooms(log_path, rooms_json_path, output_path=None):
    with open(rooms_json_path, 'r', encoding='utf-8') as f:
        rooms = json.load(f)

    # Index by canonical_id (unique hash of title+desc)
    rooms_by_canonical = {r['canonical_id']: r for r in rooms}
    rooms_by_id = {r['id']: r for r in rooms}

    last_room_id = None  # ID of the previous room we finalized
    pending_move = None  # The movement command between last room and current room
    linked_pairs = set()
    
    # Track current room to build canonical_id
    current_title = None
    current_desc = []

    def compute_canonical_id(title, desc_lines):
        """Compute canonical_id for a room"""
        desc_text = ' '.join(desc_lines) if desc_lines else ''
        
        # Strip "You also see..." from description to match parser behavior
        desc_text = re.sub(r'\.\s+You also see\s+.+?\.?\s*$', '.', desc_text, flags=re.IGNORECASE)
        
        # Match the canonical_id generation from parser (sha1, title|desc)
        h = hashlib.sha1()
        h.update(title.strip().encode('utf-8'))
        dnorm = re.sub(r'\s+', ' ', desc_text.strip())
        h.update(b"|")
        h.update(dnorm.encode('utf-8'))
        canonical_hash = h.hexdigest()[:16]
        # Build slug from title
        slug = title.lower().replace(' ', '_').replace(',', '').replace('.', '').replace("'", '')
        return f"{slug}_{canonical_hash}"

    def finalize_and_link_room():
        """Finalize current room and link it to the previous room if there was a movement"""
        nonlocal last_room_id, pending_move, current_title, current_desc
        
        if not current_title:
            return None
        
        # Compute this room's canonical_id
        canonical_id = compute_canonical_id(current_title, current_desc)
        current_room = rooms_by_canonical.get(canonical_id)
        
        if not current_room:
            # Room not in our parsed data, skip
            current_title = None
            current_desc = []
            return None
        
        # If there was a previous room and a movement between them, link them
        if last_room_id and pending_move:
            prev_room = rooms_by_id.get(last_room_id)
            if prev_room:
                rev = REVERSE.get(pending_move)
                
                # Forward link: prev_room -[direction]-> current_room
                prev_room['exits'][pending_move] = current_room['id']
                
                # Backward link: current_room -[reverse]-> prev_room  
                if rev:
                    current_room['exits'][rev] = prev_room['id']
                
                linked_pairs.add((prev_room['id'], current_room['id'], pending_move))
                
            # Clear pending_move since we've used it for linking
            pending_move = None
        
        # This room becomes the "last room" for the next iteration
        last_room_id = current_room['id']
        
        # Reset current room state
        current_title = None
        current_desc = []
        
        return current_room['id']

    with open(log_path, 'r', encoding='utf-8', errors='ignore') as f:
        for raw in f:
            line = raw.strip()
            
            if not line:
                continue

            # Detect new room header - finalize previous room and link
            h = HEADER_RE.match(line)
            if h:
                # Finalize and link the previous room if we have one
                if current_title:
                    finalize_and_link_room()
                
                # Start collecting the new room
                current_title = h.group('title')
                current_desc = []
                continue

            # Detect movement command
            m = re.match(r'^>\s*([a-z]+)', line.lower())
            if m:
                direction = m.group(1)
                if direction in ORDINALS:
                    # Normalize abbreviations to full direction names
                    direction = ABBREVIATIONS.get(direction, direction)
                    pending_move = direction
                    continue
            
            # Collect description lines (skip "Also here:" and "Obvious" lines)
            if current_title:
                if line.startswith('Also here:') or line.startswith('Obvious'):
                    continue
                if not line.startswith('>'):
                    current_desc.append(line)
    
    # Finalize the last room in the file
    if current_title:
        finalize_and_link_room()

    out_path = output_path or rooms_json_path.replace(".json", "_linked.json")
    with open(out_path, 'w', encoding='utf-8') as f:
        json.dump(rooms, f, indent=2, ensure_ascii=False)
    print(f"Linked {len(linked_pairs)} room transitions → {out_path}")

def main():
    ap = argparse.ArgumentParser(description="Link room JSONs by movement commands in log.")
    ap.add_argument("log", help="Path to the original log file (txt)")
    ap.add_argument("rooms_json", help="Path to the parsed rooms JSON")
    ap.add_argument("-o", "--output", help="Output path (default: *_linked.json)")
    args = ap.parse_args()
    link_rooms(args.log, args.rooms_json, args.output)

if __name__ == "__main__":
    main()

File: mapping/legacy/link_rooms.js
#!/usr/bin/env node
'use strict';

/**
 * link_rooms.js
 * Links parsed room JSONs by analyzing movement commands in the original log file
 * 
 * Usage: node link_rooms.js <log_file> <rooms_json_file>
 */

const fs = require('fs');
const crypto = require('crypto');

// Direction mappings
const ORDINALS = ['north', 'south', 'east', 'west', 'northeast', 'southeast', 'northwest', 'southwest', 'up', 'down', 'out'];
const ABBREVIATIONS = {
  n: 'north', s: 'south', e: 'east', w: 'west',
  ne: 'northeast', se: 'southeast', nw: 'northwest', sw: 'southwest',
  u: 'up', d: 'down'
};
const REVERSE = {
  north: 'south', south: 'north',
  east: 'west', west: 'east',
  northeast: 'southwest', southwest: 'northeast',
  northwest: 'southeast', southeast: 'northwest',
  up: 'down', down: 'up',
  out: 'out'
};

// Regex patterns
const HEADER_RE = /^\[(.+?)\]\s*$/;
const MOVEMENT_RE = /^>\s*([a-z]+)/i;

/**
 * Compute canonical_id for a room (must match parser's logic)
 */
function computeCanonicalId(title, description) {
  // Strip "You also see..." from description to match parser
  let desc = description.replace(/\.\s+You also see\s+.+?\.?\s*$/i, '.');
  
  // Normalize whitespace
  desc = desc.replace(/\s+/g, ' ').trim();
  
  // SHA1 hash of "title|description"
  const hash = crypto.createHash('sha1');
  hash.update(title.trim());
  hash.update('|');
  hash.update(desc);
  const canonicalHash = hash.digest('hex').substring(0, 16);
  
  // Create slug from title
  const slug = title.toLowerCase()
    .replace(/['']/g, '')
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '')
    .substring(0, 80);
  
  return `${slug}_${canonicalHash}`;
}

/**
 * Main linking function
 */
function linkRooms(logPath, roomsJsonPath) {
  console.log(`Reading rooms from ${roomsJsonPath}...`);
  const rooms = JSON.parse(fs.readFileSync(roomsJsonPath, 'utf8'));
  
  // Index rooms by canonical_id and by id
  const roomsByCanonical = {};
  const roomsById = {};
  for (const room of rooms) {
    roomsByCanonical[room.canonical_id] = room;
    roomsById[room.id] = room;
  }
  
  console.log(`Processing log file ${logPath}...`);
  const logContent = fs.readFileSync(logPath, 'utf8');
  const lines = logContent.split('\n');
  
  // State tracking
  let currentTitle = null;
  let currentDesc = [];
  let currentMove = null;  // Movement FROM current room
  let lastRoomId = null;   // ID of the last finalized room
  let lastMove = null;     // Movement FROM the last room
  let linkedCount = 0;
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    
    // Check for room header [Room Title]
    const headerMatch = line.match(HEADER_RE);
    if (headerMatch) {
      // Finalize the previous room if we have one
      if (currentTitle) {
        const canonicalId = computeCanonicalId(currentTitle, currentDesc.join(' '));
        const room = roomsByCanonical[canonicalId];
        
        if (room) {
          // If we have a previous room and it had a movement, link them
          if (lastRoomId && lastMove) {
            const prevRoom = roomsById[lastRoomId];
            if (prevRoom) {
              const reverse = REVERSE[lastMove];
              
              // Forward link: prevRoom --[direction]--> currentRoom
              prevRoom.exits[lastMove] = room.id;
              
              // Backward link: currentRoom --[reverse]--> prevRoom
              if (reverse) {
                room.exits[reverse] = prevRoom.id;
              }
              
              linkedCount++;
              console.log(`  Linked: ${prevRoom.id} --[${lastMove}]--> ${room.id}`);
            }
          }
          
          // This room becomes the "last room" with its movement for next iteration
          lastRoomId = room.id;
          lastMove = currentMove;
        }
      }
      
      // Start new room
      currentTitle = headerMatch[1];
      currentDesc = [];
      currentMove = null;
      continue;
    }
    
    // Check for movement command >direction
    const moveMatch = line.match(MOVEMENT_RE);
    if (moveMatch) {
      let direction = moveMatch[1].toLowerCase();
      
      // Normalize abbreviations
      direction = ABBREVIATIONS[direction] || direction;
      
      // Only record valid ordinal directions
      if (ORDINALS.includes(direction)) {
        currentMove = direction;
        console.log(`  Movement: >${direction}`);
      }
      continue;
    }
    
    // Collect description lines (skip "Obvious" and "Also here:")
    if (currentTitle && !line.startsWith('Obvious') && !line.startsWith('Also here:') && !line.startsWith('>')) {
      currentDesc.push(line);
    }
  }
  
  // Finalize the last room
  if (currentTitle) {
    const canonicalId = computeCanonicalId(currentTitle, currentDesc.join(' '));
    const room = roomsByCanonical[canonicalId];
    
    if (room && lastRoomId && lastMove) {
      const prevRoom = roomsById[lastRoomId];
      if (prevRoom) {
        const reverse = REVERSE[lastMove];
        prevRoom.exits[lastMove] = room.id;
        if (reverse) {
          room.exits[reverse] = prevRoom.id;
        }
        linkedCount++;
        console.log(`  Linked: ${prevRoom.id} --[${lastMove}]--> ${room.id}`);
      }
    }
  }
  
  // Write output
  const outputPath = roomsJsonPath.replace('.json', '_linked.json');
  fs.writeFileSync(outputPath, JSON.stringify(rooms, null, 2), 'utf8');
  
  console.log(`\nLinked ${linkedCount} room transitions → ${outputPath}`);
  
  // Calculate exit statistics
  let totalExits = 0;
  let unknownExits = 0;
  let linkedExits = 0;
  let roomsWithAllUnknown = 0;
  let fullyConnectedRooms = 0;
  
  for (const room of rooms) {
    const exitCount = Object.keys(room.exits).length;
    const unknownCount = Object.values(room.exits).filter(target => {
      if (typeof target === 'string' && target === 'unknown') return true;
      if (typeof target === 'object' && (!target.roomId || target.roomId === 'unknown')) return true;
      return false;
    }).length;
    const linkedCount = exitCount - unknownCount;
    
    totalExits += exitCount;
    unknownExits += unknownCount;
    linkedExits += linkedCount;
    
    if (exitCount > 0 && unknownCount === exitCount) {
      roomsWithAllUnknown++;
    }
    if (exitCount > 0 && unknownCount === 0) {
      fullyConnectedRooms++;
    }
  }
  
  console.log(`\n📊 Exit Statistics:`);
  console.log(`   Total rooms: ${rooms.length}`);
  console.log(`   Total exits: ${totalExits}`);
  console.log(`     Linked: ${linkedExits} (${((linkedExits/totalExits)*100).toFixed(1)}%)`);
  console.log(`     Unknown: ${unknownExits} (${((unknownExits/totalExits)*100).toFixed(1)}%)`);
  console.log(`   Rooms with all exits unknown: ${roomsWithAllUnknown}`);
  console.log(`   Fully connected rooms: ${fullyConnectedRooms}`);
}

// CLI handling
if (require.main === module) {
  const args = process.argv.slice(2);
  
  if (args.length < 2) {
    console.error('Usage: node link_rooms.js <log_file> <rooms_json_file>');
    console.error('Example: node link_rooms.js small-wl-gates-log.txt small-wl-gates-log.rooms.json');
    process.exit(1);
  }
  
  const [logFile, roomsFile] = args;
  
  try {
    linkRooms(logFile, roomsFile);
  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

module.exports = { linkRooms };

